elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,405 @@
+
+          // Set the `aria-label` to the first bit of text. Otherwise the label
+          // for this element will be computed based on its text content plus
+          // the text content of all of its children. Making it so verbose as to
+          // be unusable. This first child should ideally be a `<span>` element,
+          // but it may also just be a line of text.
+          if (!this.hasAttribute('aria-label'))
+            this.setAttribute('aria-label', this.firstChild.textContent.trim());
+      }
+    }
+  }
+
+  /**
+   * Define a custom element, `<dash-treeitem>`, and associate it with the
+   * `DashTreeItem` class.
+   */
+  window.customElements.define('dash-treeitem', DashTreeItem);
+
+  class DashTreeGroup extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      this.setAttribute('role', 'group');
+    }
+  }
+
+  /**
+   * Define a custom element, `<dash-treegroup>`, and associate it with the
+   * `DashTreeGroup` class.
+   */
+  window.customElements.define('dash-treegroup', DashTreeGroup);
+
+  class DashTree extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'tree');
+      this.setAttribute('tabindex', 0);
+
+      // Before the elements starts booting, it waits for both
+      // the `<dash-treeitem>` and `<dash-treegroup>` to load.
+      Promise.all([
+        customElements.whenDefined('dash-treeitem'),
+        customElements.whenDefined('dash-treegroup'),
+      ]).then(_ => {
+        // Acquire all treeitems inside the element
+        const treeItems = this._allTreeItems();
+        // If there are no treeItems, then the tree is empty. Abort.
+        if (treeItems.length === 0) return;
+
+        // TODO: Handle a child that is already aria-selected=true?
+      });
+    }
+
+    _allTreeItems() {
+      const treeItems = [];
+      // A recursive function that visits every child and builds a tree
+      // This produces the same results as calling querySelectorAll,
+      // but writing the tree ourselves lets us toggle which elements
+      // are visible.
+      function findTreeitems(node, isVisible) {
+        let el = node.firstElementChild;
+        while (el) {
+          // Ignore children like `<span>` and `<dash-treegroup>`.
+          // If the element has children, descend into them looking for
+          // more treeitems.
+          if (!isTreeItem(el)) {
+            if (el.firstElementChild) {
+              findTreeitems(el, isVisible);
+              return;
+            }
+          }
+
+          if (isTreeItem(el)) {
+            treeItems.push(el);
+            el.visible = isVisible;
```

Comment:
nevermind, I figured out how to do this with a simpler tree walk. Basically as soon as I hit an item that's not visible I stop walking the tree in that direction. This simplifies going next/prev as well since invisible nodes aren't even in the tree.